Refactoring to including ros2 lib build (#7)
rapyuta-robotics/UE_tools: README.md

    This script will create **ros2_ws** under BuildROS2 and build there.
    
    \* This script will modify local ros2 installation. Need to `sudo apt install ros-foxy-desktop` if you want to use standard ros2 foxy. todo: build inside docker to avoid affecting local setup.
    

    \* **Note**
    
    - Reinstall python package due to issues https://github.com/ros-visualization/qt_gui_core/issues/212:

- **build_and_install_ros2_pkgs.py**: 
    
    Build ros2 foxy package inside **ros2_ws** which is created by `build_and_install_ros2_pkgs.py` and copy lib and header files to UnrealEngine project folder. This is mainly used to add/update new msgs to UnrealEngine project.

    This script will get source listed in **ros2_additional_pkgs.repos** and build pkgs listed in top of the script or given arg.

    You need to follow CodeGen tutorial as well to use msg from UnrealEngine.

## CodeGen
Python script to generate UE4 .h and .cpp files for UnrealEngine to interface with ROS2 messages.

> \* Based on Jinja2 to generatee msg

Reviewer: ```suggestion
\* Based on Jinja2 to generate msg
```

\* Ignores message types deprecated in Foxy.

### Usage

#### Generate source files
1. cd UE_tools/CodeGen
2. python3 gen_ue_from_ros.py --module <UE module name used in class/struct definition> --dependency <path to ros2 dependency pkgs> --target <names of target pkgs>

    e.g.

    `python3 gen_ue_from_ros.py`

    \* default dependency and target pkgs are defined at top of gen_ue_from_ros.py

#### Copy source files to Unreal Project
1. cd UE_tools/CodeGen
2. python3 post_genrate.py --ue_proj_path <path to UnrealEngine project> 

    e.g.

    `python3 post_generate.py --ue_proj_path /home/user/turtlebot3-UE`

### Scripts
- **gen_ue_from_ros.py**: 

    Generate .h and cpp files for UnrealEngine project from msg/srv/action definitions. Generated class/struct can be used from UnrealEngine C++ and Blueprint. 
    

>     All properties can be accessible from C++. Not all properties but only Blueprint comaptible type member can be accesible from  Blueprint.

Reviewer: ```suggestion
    All properties can be accessible from C++. Not all properties but only Blueprint compatible type member can be accessible from Blueprint.
```

>     If your custom msg need other msg dependency, you need to specify that dependnecy from arg.

Reviewer: ```suggestion
    If your custom msg need other msg dependency, you need to specify that dependency from arg.
```

    \* Default dependencies and targets are defined near the top of files.

>     \* Please check [limitiations](#limitation) as well.

Reviewer: ```suggestion
    \* Please check [limitations](#limitation) as well.
```

- **post_generate.py**: 

    Copy generated .h and cpp to target UnrealEngine project folder such as **turtlebot3-UE/Plugin/rclUE/Source/Thirdparty/ros**. 

>     \* Black list is deifined near the beginning of the script which are not copied to Unreal project since there are [Limitiations](#limitation)

Reviewer: ```suggestion
    \* Black list is defined near the beginning of the script which are not copied to Unreal project since there are [Limitations](#limitation)
```

### Limitation
- only works with ROS2 message interface (in particular, ROS had built-in data types, such as `time`, defined in libraries and ROS2 now implements those as messages)
- code generation for nested arrays in messages is not supported
- currently it has only been tested with messages used in RR projects
- not all types are supported in UE4 Blueprint (e.g. `double`): `get_types_cpp` does the check, however it is currently checking against a list of unsupported types that have been encountered (and there's more that are not checked against, so if the code fails compilation due to this problem, the type in question should be currently be added to the list). The alternative, and better implementation, would check for supported types (but must be careful with various aliases, like `int` and `int32`

> - fixed size array comes with TArray<>. User should researve array with proper size by themselves.

Reviewer: ```suggestion
- fixed size array comes with TArray<>. User should reserve array with proper size by themselves.
```
